### PR TITLE
Improve writeup on recursion

### PR DESCRIPTION
**Pull request by kxmbrian**

### CODE REVIEW

File: chapters/3-recursion.md [hunk 1]
```diff
@@ -1,8 +1,147 @@
...
+
+To resolve this, we can introduce the idea of an execution *frame*. It becomes clear that Line 5 is being run for the  `factorial(3)` frame and Line 2 is being run for the `factorial(2)` frame. Right now, the frame for `factorial(2)` is the current (active) frame while the rest are not currently being run.
+
+Another way to work around the confusion is to imagine that a new copy of the code is created every time a recursive call is made. When there are multiple frames stuck on the same line, we can easily visualise which they are.
+
+#### Wishful Thinking
+
+Writing recursive functions require you to take a step of faith.
+
+One day, Leonardo was doing his programming assignment. His teacher had tasked the class to write `fibonnaci`. He thought he had his answer nailed down and proceeded to type his solution out:
```
Comment (fonglh) [inline chapters/3-recursion.md]: fibonacci. single n double c

File: chapters/3-recursion.md [hunk 2]
```diff
@@ -1,8 +1,147 @@
...
+"Before the code can be completed on your machine, it must be complete in your mind. Have faith that you will be able to correctly implement the function that you are about to use. Only if you take the leap and use that function, will that function ever be correct. I believe in you."
+
+Leonardo took his teacher's instruction to heart and with wanton *wishful thinking*, simply (and almost carelessly) assumed that any function that he wanted to use existed. He used them first, worrying about their existence later.
+
+Armed with this insight, he went on to have an illustrious career as a computer programmer. His crowing achievement was to discover the [Fibonnaci](https://en.wikipedia.org/wiki/Fibonacci) numbers. After many years, he decided that he wanted to raise the next generation and became a teacher.
+
+He was exactly his teacher's age when he started his new career. The first assignment that he set for his students was to write a function to generate his namesake numbers. There happened to be in his class a bright boy named Leonardo.
+
```
Comment (fonglh) [inline chapters/3-recursion.md:143]: This is a bit freaky, haha

**Comment by fonglh** [comment]
Ok, just need to fix the typos

**Comment by kxmbrian** [comment]
Ok, fixed

**Pull request merged**